R/tree.R

```diff
@@ -0,0 +1,386 @@
+  show_environments = TRUE,
+  val_printer = crayon::blue,
+  class_printer = crayon::silver,
+  show_attributes = FALSE,
+  remove_newlines = TRUE,
+  tree_chars = box_chars()
+){
+  ellipsis::check_dots_empty()
+
+  # Pack up the unchanging arguments into a list and send to tree_internal
+  termination_type <- tree_internal(
+    x,
+    opts = list(
+      index_unnamed = index_unnamed,
+      max_depth = max_depth,
+      max_length = max_length,
+      show_envs = show_environments,
+      val_printer = val_printer,
+      class_printer = class_printer,
+      show_attributes = show_attributes,
+      remove_newlines = remove_newlines,
+      tree_chars = tree_chars
+    )
+  )
+  if (termination_type == "early") {
+    cat("...", "\n")
+  }
+
+  invisible(x)
+}
+
+# Tree printing internal function
+#
+# This is the internal function for the main tree printing code. It wraps the
+# static options arguments from the user-facing `tree()` into a single opts
+# list to make recursive calls cleaner. It also has arguments that as it is
+# called successively but the end-user shouldn't see or use.
+tree_internal <- function(
+  x,
+  x_id = NULL,
+  branch_hist = character(0),
+  opts,
+  attr_mode = FALSE,
+  counter_env = rlang::new_environment(data = list(n_elements_printed = 0, envs_seen = c()))
+) {
+  counter_env$n_elements_printed <- counter_env$n_elements_printed + 1
+  # Stop if we've reached the max number of times printed desired
+  if (counter_env$n_elements_printed > opts$max_length) {
+    return("early")
+  }
+
+  depth <- length(branch_hist)
+
+  # Build branch string from branch history
+  # Start with empty spaces
+  branch_chars <- rep_len("  ", depth)
+
+  branch_chars[branch_hist == "child"] <- paste0(opts$tree_chars$v, " ")
+  branch_chars[branch_hist == "pre-attrs"] <- paste0(opts$tree_chars$vd, " ")
+
+  # Next update the final element (aka the current step) with the correct branch type
+  last_step <- branch_hist[depth]
+  root_node <- length(branch_hist) == 0
+
+  branch_chars[depth] <- if (root_node) {
+    ""
+  }  else {
+    paste0(
+      if (grepl("last", last_step)) opts$tree_chars$l else opts$tree_chars$j,
+      if (grepl("attribute", last_step)) opts$tree_chars$hd else opts$tree_chars$h
+    )
+  }
+
+  # Build label
+  label <- paste0(
+    x_id,
+    if (!rlang::is_null(x_id) && x_id != "") ": ",
+    tree_label(
+      x,
+      class_printer = opts$class_printer,
+      val_printer = opts$val_printer,
+      remove_newlines = opts$remove_newlines
+    )
+  )
+
+  # Figure out how many children we have (plus attributes if they are being
+  # printed) so we can setup how to proceed
+  x_attributes <- attributes(x)
+  if (attr_mode){
+    # Filter out "names" attribute as this is already shown by tree
+    x_attributes <- x_attributes[names(x_attributes) != "names"]
+  }
+  has_attributes <- length(x_attributes) > 0 & opts$show_attributes
+
+  has_children <- has_attributes | length(x) > 1
```

Comment:
Better to use `||` here and elsewhere, since these are always supposed to be a single `TRUE` or `FALSE`